explorer/interpreter/impl_scope.cpp

```diff
@@ -137,6 +141,41 @@ auto ImplScope::ResolveInterface(Nonnull<const InterfaceType*> iface_type,
   return *result;
 }
 
+// Combines the results of two impl lookups.
+static auto CombineResults(Nonnull<const InterfaceType*> iface_type,
+                           Nonnull<const Value*> type,
+                           SourceLocation source_loc,
+                           std::optional<Nonnull<const Witness*>> a,
+                           std::optional<Nonnull<const Witness*>> b)
+    -> ErrorOr<std::optional<Nonnull<const Witness*>>> {
+  // If only one lookup succeeded, return that.
+  if (!a) {
+    return b;
+  }
+  if (!b) {
+    return a;
+  }
+  // If either of them was a symbolic result, then they'll end up being
+  // equivalent. Pick whichever we found first.
+  if (!isa<ImplWitness>(*b)) {
+    return a;
+  }
+  if (!isa<ImplWitness>(*a)) {
+    return b;
+  }
+  // If they refer to the same `impl` declaration, it doesn't matter which one
+  // we pick.
+  auto* impl_a = cast<ImplWitness>(*a);
+  auto* impl_b = cast<ImplWitness>(*b);
```

Comment:
```suggestion
  auto* impl_b = dyn_cast<ImplWitness>(*b);
  if (impl_b == nullptr) {
    return a;
  }
  auto* impl_a = dyn_cast<ImplWitness>(*a);
  if (impl_a == nullptr) {
    return b;
  }
  // If they refer to the same `impl` declaration, it doesn't matter which one
  // we pick.
```

Can isa+cast be combined into dyn_cast?

But also, I'm really tempted to suggest combining further, like:

```
  auto* impl_b = b ? nullptr : dyn_cast<ImplWitness>(*b);
  if (impl_b == nullptr) {
    return a;
  }
```

Reply:
Combined `isa` + `cast`. I think the further combination doesn't work, because it would mean that merging a witness that's not an `ImplWitness` with a `std::nullopt` would return `std::nullopt`, which is not the intended semantics.